### ISSUE TITLE
New "Mesh to Curves" and/or "Bezier in Lite" nodes"

### ISSUE THREAD

**Issue by maxim1967** [nortikin/sverchok#4047]
**A suggestion for creating new nodes.**
New "Mesh to Curves" and/or "Bezier in Lite" nodes. The "Mesh to Curves" node-converts a mesh curve to a bezier curve. The node "Bezier in Lite" is similar in functionality to " Objects in Lite"

**Comment by maxim1967** [comment]
The "Bezier in Lite" node is more preferable, and there is less chance of geometry distortion during conversion.

**Comment by zeffii** [comment]
bezier_in  isn't a fantastically complicated node, it might be preferable to store the bezier control data inside  the current node, rather than making a Lite version. which would duplicate everything and then add the storage part.

**Comment by maxim1967** [comment]
No, it should be two different nodes "Bezier in" is a node for working with curves on the fly, its main task is to use curves in the scene, and "Bezier in Lite" is primarily intended for storing saved curves. For example, the discussed addition of "To Control Panel" for "Bezier in Lite" will be superfluous.

**Comment by maxim1967** [comment]
This is also how we have "Objects in" and "Objects in Lite". In addition, "Objects in" can be upgraded, improved, and added new functionality that is not needed in "Objects in Lite".

**Comment by rendetto** [comment]
What is a "mesh curve"?

**Comment by maxim1967** [comment]
Strange question. Moreover, it says " Mesh to Curves"

**Comment by rendetto** [comment]
I asked because if you refer to a polyedge (or just free vertices) to a curve conversion there are several interpolation and approximation nodes. I was wondering if the "mesh curve" you mentioned is something else - for example a subdivison based curve (I was using those a lot back in the days because the NURBS support in Blender is so poor and lacks snapping).
Thаt is what I meant:
![subd_curve](https://user-images.githubusercontent.com/66558924/114917368-88753900-9e2e-11eb-9f6e-fb9bf5d9bedc.jpg)

**Comment by maxim1967** [comment]
...converts a mesh curve to a bezier curve. Analogous to the Blender menu item "Convert to" - " Curve"

**Comment by maxim1967** [comment]
Good people, how about "Bezier in Lite" there is a hope that this node will appear?

**Comment by portnov** [comment]
As far as I understand, you actually want two flags for "Bezier In" node:
* [x] Store curve with node, when exporting to JSON
* [ ] When importing from JSON, load curve as object into scene.

Correct?

**Comment by maxim1967** [comment]
When importing from JSON, load the curve as a bezier curve object into the scene.

**Comment by maxim1967** [comment]
Here I tried to create what I wanted:

`import bpy
from bpy.props import FloatProperty, EnumProperty, BoolProperty, StringProperty
from mathutils import Vector

from sverchok.node_tree import SverchCustomTreeNode
from sverchok.utils.nodes_mixins.sv_animatable_nodes import SvAnimatableNode
from sverchok.utils.nodes_mixins.show_3d_properties import Show3DProperties
from sverchok.utils.sv_operator_utils import SvGenericNodeLocator
from sverchok.data_structure import updateNode, zip_long_repeat, split_by_count
from sverchok.utils.curve.algorithms import concatenate_curves
from sverchok.utils.curve.bezier import SvCubicBezierCurve


class SvBezierInLiteCallbackOp(bpy.types.Operator, SvGenericNodeLocator):

    bl_idname = "node.sv_bezier_in_lite_callback"
    bl_label = "Bezier In Lite Callback"
    bl_options = {'INTERNAL'}

    def execute(self, context):
        """
        returns the operator's 'self' too to allow the code being called to
        print from self.report.
        """
        node = self.get_node(context)
        if node:
            node.get_objects_from_scene(self)
            return {'FINISHED'}

        return {'CANCELLED'}


class SvBezierInLiteNode(Show3DProperties, bpy.types.Node, SverchCustomTreeNode, SvAnimatableNode):
    """
    Triggers: Input Bezier
    Tooltip: Get Bezier Curve objects from scene
    """
    bl_idname = 'SvBezierInLiteNode'
    bl_label = 'Bezier In Lite'
    bl_icon = 'OUTLINER_OB_EMPTY'
    sv_icon = 'SV_OBJECTS_IN'

    object_names: bpy.props.CollectionProperty(type=bpy.types.PropertyGroup)

    sort: BoolProperty(
        name='sort by name',
        description='sorting inserted objects by names',
        default=True, update=updateNode)

    apply_matrix: BoolProperty(
        name = "Apply matrices",
        description = "Apply object matrices to control points",
        default = True,
        update = updateNode)
    node_dict = {}


    def sv_init(self, context):
        self.outputs.new('SvCurveSocket', 'Curves')
        self.outputs.new('SvVerticesSocket', 'ControlPoints')
        self.outputs.new('SvMatrixSocket', 'Matrices')

    def get_objects_from_scene(self, ops):
        """
        Collect selected objects
        """
        self.object_names.clear()

        names = [obj.name for obj in bpy.data.objects if (obj.select_get() and len(obj.users_scene) > 0 and len(obj.users_collection) > 0)]

        if self.sort:
            names.sort()

        for name in names:
            self.object_names.add().name = name

        if not self.object_names:
            ops.report({'WARNING'}, "Warning, no selected objects in the scene")
            return

        self.process_node(None)

    def draw_obj_names(self, layout):
        # display names currently being tracked, stop at the first 5..
        if self.object_names:
            remain = len(self.object_names) - 5

            for i, obj_ref in enumerate(self.object_names):
                layout.label(text=obj_ref.name)
                if i > 4 and remain > 0:
                    postfix = ('' if remain == 1 else 's')
                    more_items = '... {0} more item' + postfix
                    layout.label(text=more_items.format(remain))
                    break
        else:
            layout.label(text='--None--')

    def draw_buttons_ext(self, context, layout):
        layout.prop(self, "draw_3dpanel")

    def draw_buttons_3dpanel(self, layout):
        row = layout.row(align=True)
        row.label(text=self.label if self.label else self.name)
        self.wrapper_tracked_ui_draw_op(row, SvBezierInLiteCallbackOp.bl_idname, text='GET')
        self.wrapper_tracked_ui_draw_op(row, "node.sv_nodeview_zoom_border", text="", icon="TRACKER_DATA")

    def draw_buttons(self, context, layout):
        self.draw_animatable_buttons(layout, icon_only=True)
        col = layout.column(align=True)
        row = col.row(align=True)

        row = col.row()
        op_text = "Get selection"  # fallback

        try:
            addon = context.preferences.addons.get(sverchok.__name__)
            if addon.preferences.over_sized_buttons:
                row.scale_y = 4.0
                op_text = "G E T"
        except:
            pass

        self.wrapper_tracked_ui_draw_op(row, SvBezierInLiteCallbackOp.bl_idname, text=op_text)

        layout.prop(self, 'sort', text='Sort', toggle=True)
        layout.prop(self, 'apply_matrix', toggle=True)

        self.draw_obj_names(layout)

    def get_curve(self, spline, matrix):
        segments = []
        pairs = zip(spline.bezier_points, spline.bezier_points[1:])
        if spline.use_cyclic_u:
            pairs = list(pairs) + [(spline.bezier_points[-1], spline.bezier_points[0])]
        points = []
        is_first = True
        for p1, p2 in pairs:
            c0 = p1.co
            c1 = p1.handle_right
            c2 = p2.handle_left
            c3 = p2.co
            if self.apply_matrix:
                c0, c1, c2, c3 = [tuple(matrix @ c) for c in [c0, c1, c2, c3]]
            else:
                c0, c1, c2, c3 = [tuple(c) for c in [c0, c1, c2, c3]]
            points.append([c0, c1, c2, c3])
            segment = SvCubicBezierCurve(c0, c1, c2, c3)
            segments.append(segment)
        return points, concatenate_curves(segments)

    def process(self):

        if not self.object_names:
            return

        curves_out = []
        matrices_out = []
        controls_out = []
        for item in self.object_names:
            object_name = item.name
            obj = bpy.data.objects.get(object_name)
            if not obj:
                continue
            with self.sv_throttle_tree_update():
                matrix = obj.matrix_world
                if obj.type != 'CURVE':
                    self.warning("%s: not supported object type: %s", object_name, obj.type)
                    continue
                for spline in obj.data.splines:
                    if spline.type != 'BEZIER':
                        self.warning("%s: not supported spline type: %s", spline, spline.type)
                        continue
                    controls, curve = self.get_curve(spline, matrix)
                    curves_out.append(curve)
                    controls_out.append(controls)
                    matrices_out.append(matrix)

        self.outputs['Curves'].sv_set(curves_out)
        self.outputs['ControlPoints'].sv_set(controls_out)
        self.outputs['Matrices'].sv_set(matrices_out)
    
    def load_from_json(self, node_data: dict, import_version: float):
        if 'curv' not in node_data:
            return  # looks like a node was empty when it was imported
        curv = node_data['curv']
        name = node_data['params']["obj_name"]
        curv_dict = json.loads(curv)

        if not curv_dict:
            print(self.name, 'contains no flatten curv')
            return

        unrolled_curv = unflatten(curv_dict)
        verts = unrolled_curv['Curves']
        controlpoints = unrolled_curv['ControlPoints']
        matrix = unrolled_curv['Matrices']

        if self.do_not_add_obj_to_scene:
            self.node_dict[hash(self)] = unrolled_curv
            self.obj_name = name
            return
            

    def save_to_json(self, node_data: dict):
        # generate flat data, and inject into incoming storage variable
        obj = self.node_dict.get(hash(self))
        print(obj)
        if not obj:
            self.error('failed to obtain local geometry, can not add to json')
            return

        node_data['curv'] = json.dumps(flatten(obj))


def register():
    bpy.utils.register_class(SvBezierInLiteCallbackOp)
    bpy.utils.register_class(SvBezierInLiteNode)

def unregister():
    bpy.utils.unregister_class(SvBezierInLiteNode)
    bpy.utils.unregister_class(SvBezierInLiteCallbackOp)
`

That's just it doesn't work can you tell me what's wrong?

**Comment by Durman** [comment]
Is there any error or what?

**Comment by maxim1967** [comment]
This node is not registered, it is not in the list of nodes and why this happens I do not understand.

**Comment by Durman** [comment]
You also have to add the name of your node class into this file `sverchok/index.md`

**Comment by maxim1967** [comment]
Durman / Thanks for the help. Here's a little tinkering with the code:

```
import bpy
from bpy.props import FloatProperty, EnumProperty, BoolProperty, StringProperty
from mathutils import Vector

from sverchok.node_tree import SverchCustomTreeNode
from sverchok.utils.nodes_mixins.sv_animatable_nodes import SvAnimatableNode
from sverchok.utils.nodes_mixins.show_3d_properties import Show3DProperties
from sverchok.utils.sv_operator_utils import SvGenericNodeLocator
from sverchok.data_structure import updateNode, zip_long_repeat, split_by_count
from sverchok.utils.curve.algorithms import concatenate_curves
from sverchok.utils.curve.bezier import SvCubicBezierCurve

import json

class SvBezierInLiteCallbackOp(bpy.types.Operator, SvGenericNodeLocator):

    bl_idname = "node.sv_bezier_in_lite_callback"
    bl_label = "Bezier In Lite Callback"
    bl_options = {'REGISTER', 'UNDO', 'INTERNAL'}
    

    def execute(self, context):
        """
        returns the operator's 'self' too to allow the code being called to
        print from self.report.
        """
        node = self.get_node(context)
        if node:
            node.get_objects_from_scene(self)
            return {'FINISHED'}

        return {'CANCELLED'}
        
class SvBezierInLiteNode(Show3DProperties, bpy.types.Node, SverchCustomTreeNode, SvAnimatableNode):
    """
    Triggers: Input Bezier
    Tooltip: Get Bezier Curve objects from scene
    """
    bl_idname = 'SvBezierInLiteNode'
    bl_label = 'Bezier In Lite'
    bl_icon = 'OUTLINER_OB_EMPTY'
    sv_icon = 'SV_OBJECTS_IN'
    
    do_not_add_obj_to_scene: BoolProperty(
        default=False,
        description="Do not add the object to the scene if this node is imported from elsewhere")
        
    apply_matrix: BoolProperty(
        name = "Apply matrices",
        description = "Apply object matrices to control points",
        default = True,
        update = updateNode)

    currently_storing: BoolProperty()
    obj_name: bpy.props.CollectionProperty(type=bpy.types.PropertyGroup)
    node_dict = {}
    
    def get_objects_from_scene(self, ops):
        """
        Collect selected objects
        """
        self.obj_name.clear()

        names = [obj.name for obj in bpy.data.objects if (obj.select_get() and len(obj.users_scene) > 0 and len(obj.users_collection) > 0)]

        for name in names:
            self.obj_name.add().name = name

        if not self.obj_name:
            ops.report({'WARNING'}, "Warning, no selected objects in the scene")
            return

        self.process_node(None)
    
    def drop(self):
        self.obj_name = ""
        self.currently_storing = False
        self.node_dict[hash(self)] = {}
        
    def get_curve(self, spline, matrix):
        segments = []
        pairs = zip(spline.bezier_points, spline.bezier_points[1:])
        if spline.use_cyclic_u:
            pairs = list(pairs) + [(spline.bezier_points[-1], spline.bezier_points[0])]
        points = []
        is_first = True
        for p1, p2 in pairs:
            c0 = p1.co
            c1 = p1.handle_right
            c2 = p2.handle_left
            c3 = p2.co
            if self.apply_matrix:
                c0, c1, c2, c3 = [tuple(matrix @ c) for c in [c0, c1, c2, c3]]
            else:
                c0, c1, c2, c3 = [tuple(c) for c in [c0, c1, c2, c3]]
            points.append([c0, c1, c2, c3])
            segment = SvCubicBezierCurve(c0, c1, c2, c3)
            segments.append(segment)
        return points, concatenate_curves(segments)
        
    def sv_init(self, context):
        self.outputs.new('SvCurveSocket', 'Curves')
        self.outputs.new('SvVerticesSocket', 'ControlPoints')
        self.outputs.new('SvMatrixSocket', 'Matrices')
        
    def draw_obj_names(self, layout):
        # display names currently being tracked, stop at the first 5..
        if self.obj_name:
            remain = len(self.obj_name) - 5

            for i, obj_ref in enumerate(self.obj_name):
                layout.label(text=obj_ref.name)
                if i > 4 and remain > 0:
                    postfix = ('' if remain == 1 else 's')
                    more_items = '... {0} more item' + postfix
                    layout.label(text=more_items.format(remain))
                    break
        else:
            layout.label(text='--None--')
        
    def draw_buttons(self, context, layout):
        callback = 'node.sv_bezier_in_lite_callback'

        col = layout.column(align=True)
        row = col.row(align=True)

        row = col.row()
        op_text = "Get selection"  # fallback

        try:
            addon = context.preferences.addons.get(sverchok.__name__)
            if addon.preferences.over_sized_buttons:
                row.scale_y = 4.0
                op_text = "G E T"
        except:
            pass

        self.wrapper_tracked_ui_draw_op(row, callback, text=op_text)
        
        layout.prop(self, 'apply_matrix', toggle=True)

        self.draw_obj_names(layout)
        
    def pass_data_to_sockets(self):
        curv_data = self.node_dict.get(hash(self))
        if curv_data:
            for socket in self.outputs:
                if socket.is_linked:
                    socket.sv_set([curv_data[socket.name]])
                    
    def process(self):

        if not hash(self) in self.node_dict:
            if self.obj_name and bpy.data.objects.get(self.obj_name):
                self.get_curve(self.obj_name)
            else:
                self.debug('ending early, no node_dict')
                return

        self.pass_data_to_sockets()
        
    def load_from_json(self, node_data: dict, import_version: float):
        if 'curv' not in node_data:
            return  # looks like a node was empty when it was imported
        curv = node_data['curv']
        name = node_data['params']["obj_name"]
        curv_dict = json.loads(curv)

        if not curv_dict:
            print(self.name, 'contains no flatten curv')
            return

        unrolled_curv = unflatten(curv_dict)
        verts = unrolled_curv['Curves']
        controlpoints = unrolled_curv['ControlPoints']
        matrix = unrolled_curv['Matrices']

        if self.do_not_add_obj_to_scene:
            self.node_dict[hash(self)] = unrolled_curv
            self.obj_name = name
            return
            
    def save_to_json(self, node_data: dict):
        # generate flat data, and inject into incoming storage variable
        obj = self.node_dict.get(hash(self))
        print(obj)
        if not obj:
            self.error('failed to obtain local geometry, can not add to json')
            return

        node_data['curv'] = json.dumps(flatten(obj))


def register():
    bpy.utils.register_class(SvBezierInLiteCallbackOp)
    bpy.utils.register_class(SvBezierInLiteNode)

def unregister():
    bpy.utils.unregister_class(SvBezierInLiteNode)
    bpy.utils.unregister_class(SvBezierInLiteCallbackOp)
```


I'm stuck on this.

![Снимок3](https://user-images.githubusercontent.com/62519357/116230711-a876f780-a760-11eb-8254-97c85faf763a.JPG)

I don't understand the error.

**Comment by Durman** [comment]
It's strange that the error points into the wrong line. Probably you edited the code and did not reload Sverchok.
```py
        if self.obj_name and bpy.data.objects.get(self.obj_name):
            self.get_curve(self.obj_name)
```
This strings are incorrect. `self.obj_name` is actually a collection and not a string. I think you have to iterate over all ojbect names to create curves for all objects.

**Comment by maxim1967** [comment]
Something like that:
```py
def process(self):
    
        if not hash(self) in self.node_dict:
            if not self.obj_name:
                return

            curves_out = []
            matrices_out = []
            controls_out = []
            for item in self.obj_name:
                object_name = item.name
                obj = bpy.data.objects.get(object_name)
                if not obj:
                    continue
                with self.sv_throttle_tree_update():
                    matrix = obj.matrix_world
                    if obj.type != 'CURVE':
                        self.warning("%s: not supported object type: %s", object_name, obj.type)
                        continue
                    for spline in obj.data.splines:
                        if spline.type != 'BEZIER':
                            self.warning("%s: not supported spline type: %s", spline, spline.type)
                            continue
                        controls, curve = self.get_curve(spline, matrix)
                        curves_out.append(curve)
                        controls_out.append(controls)
                        matrices_out.append(matrix)

            self.outputs['Curves'].sv_set(curves_out)
            self.outputs['ControlPoints'].sv_set(controls_out)
            self.outputs['Matrices'].sv_set(matrices_out)

        self.pass_data_to_sockets()
```
the error has been fixed. But the json file is not saved, alas.

**Comment by Durman** [comment]
Do you mean information about curves does not get into the json file during export into json?

**Comment by maxim1967** [comment]
Yes, that's exactly what I need. Save the curve object in a json file.

**Comment by Durman** [comment]
Could you share the JSON file to see whether the problem is in export or in import?

**Comment by maxim1967** [comment]
[test.zip](https://github.com/nortikin/sverchok/files/6392345/test.zip)
Most likely in export

**Comment by Durman** [comment]
Actually I can't see the source of the problem looking at your code snippets. Could you share whole code of your node or create pull request so I can test it?

**Comment by maxim1967** [comment]
Made a pull request
https://github.com/nortikin/sverchok/pull/4067
